design/build-clusters.md

```diff
@@ -0,0 +1,140 @@
+# Multiple Build Cluster 
+
+## Context
+
+Build a scalable Screwdriver build infrastructure. 
+
+## Objective 
+
+1. Running builds in parallel with help of multiple build clusters.
+2. Users can bring in their own build clusters for their own specific infrastructure needs.
+3. Support build cluster in AWS
+
+## Proposal
+
+1. Implicit build clusters
+	Screwdriver maintains its own multiple build cluster infrastructure.
+
+2. Explicit build clusters
+	Users can bring in their own build cluster infrastructure. 
+
+As part of cluster onboarding process for above both options cluster admin should register their build cluster information with Screwdriver. Below details are required to register a build cluster.
+	1. Cluster name
+	2. Status - cluster status
+	3. Queue name
+	4. Authorization (initial phase will support JWT with key and sign options)
+	5. SCM Organization
+	6. SCM Context
+	7. Managed (Screwdriver / External) 
+
+Multiple build cluster onboarding process doc 
+	TBD
+
+## Design
+
+Redis queue in buildClusters
+
+![build-clusters-design.png](diagrams/build-clusters-design.png)
+
+
+### Scheduler service will be responsible 
+	1. identify build cluster and queue information for a build and queue jobs in respective queues.
+
+	2. authorize queue worker from build cluster and allow only authorized jobs from queues wrto build cluster.
+
+### Authorization
+Initial phase, we will go with JWT + private and public key authorization. Token expiry will be passed as part of signOption. Periodically cycle private+public key and signOption which has the expiry interval, and this will be a manual step which needs to be co-ordinated between Build cluster admin and Screwdriver team.
+
+### Yaml 
+
+```yml
+shared:
+    environment:
+    NODE_ENV: test
+    settings:
+        email:
+    addresses: [test@email.com, test2@email.com]
+    statuses: [SUCCESS, FAILURE]
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
+    publish:
+    requires: main
+    image: node:6
+    steps:
+        - publish: npm publish
+    ...
+```   
+ 
+### New table for build cluster details
+Table: `buildClusters`
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganization` | text(200) | no | no | no | |
+| `scmContext` | text(200) | no | no | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
+| `authKeyFileName` | text(100) | no | no | yes | *encrypt and store the public key filename* |
```

Comment:
Why we want it to be a file in the API container? It can be an ENV to api like this: https://github.com/screwdriver-cd/screwdriver/blob/master/config/custom-environment-variables.yaml#L8

Reply:
agree, will update the doc